docs: exit game properties + tx type dependencies (#455)
omisego/plasma-contracts: plasma_framework/docs/design/payment-game-implementation-v1.md

# Payment ExitGame Implementation V1

This document is a short description of the interaction functions of the current Payment ExitGame implementation. Please check the [ALD ExitGame definitions and properties doc](./ald-exit-game-definitions-and-properties.md) for the description of each specific phrases used in here.

In Payment transaction V1, it only allows to have Payment transaction V1 as inputs and have both Payment transaction V1 and V2 as outputs. Specification of V2 is still undefined. However, the spending condition of V1 to V2 is defined as the signature of output owner should be correctly signed.

## Transaction Data Structure
```
struct Transaction {
    uint256 txType;
    bytes32[] inputs;
    Output[] outputs;
    bytes32 metaData;
}

struct Output {
    uint256 outputType;
    bytes20 outputGuard;
    address token;
    uint256 amount;
}
```

> In the first version of our Plasma Framework system, there would be Payment Transaction running only. 

Reviewer: ```suggestion
In the first version of the Plasma Framework system, there is only one type of Payment Transaction. 
```

> 1. `txType` represents the type of the transaction. Each transaction type would have one and only one exit game contract mapping the exit game interactions. 

Reviewer: ```suggestion
1. `txType` represents the type of the transaction. Each transaction type has only one exit game contract mapping the exit game interactions. 
```

> 1. The `inputs` field would be an array of `bytes32` holding `utxoPos` as the output identifier. There would be at most 4 inputs. For deposit transaction it should be with no inputs.

Reviewer: ```suggestion
1. The `inputs` field is an array of `bytes32` representing the `utxoPos` as the output identifier. There are a maximum four inputs. A deposit transaction does have zero inputs.
```

> 1. The `outputs` field would be an array of `Output` struct holding payment output information. There would be at most 4 outputs. For deposit transaction it should be with one and only one output.

Reviewer: ```suggestion
1. The `outputs` field is an array of `Output` struct representing payment output information. There are a maximum four outputs.  A deposit transaction does have one output.
```

> 1. `metaData` field is a `bytes32` that can be used to put extra data for the transaction. It should be an empty bytes if the transaction does not want to hold any data in this field.

Reviewer: ```suggestion
1. `metaData` field is a `bytes32` field that can be used to add extra data to the transaction. It should contain zeros if the transaction does not have an extra data.
```

Author: > It should contain zeros if the transaction does not have an extra data.

awwwwww....one more potential encode/decode issue. We cannot differentiate bytes with zero and without from RLP.

Reviewer 2: I think the encoded `metadata` binary field should be interpreted as `bytes32` after/during decoding, so with zero-padding. Not 100% sure though here.

> Output is of the four elements that fulfills the current `WireTransactionOutput`:

Reviewer: ```suggestion
Output has four elements that fulfills the current `WireTransactionOutput`:
```

> 1. `outputType` represents the type of the output. The output type would be used to decide, eg. spending condition. Each new valid transaction type should owns new output type.

Reviewer: ```suggestion
1. `outputType` represents the type of output. The output type would is used to decide, eg. spending condition. Each new valid transaction type is tied to a new output type.
```

Author: > Each new valid transaction type is tied to a new output type.

This is a bit off actually. It is probably more like each new valid transaction type is tied to one or more new output types.

Author: added some explanation, might want to take a look

> 1. `outputGuard` is the field that holds the authentication data of the output. In the first version of Payment transaction, it would be always holding the `owner` address of the output directly. For instance, if the output belongs to Alice, then this field would be Alice's address.

Reviewer: ```suggestion
1. `outputGuard` is the field that represents the authentication data of the output. It value must always be the same than the `owner` address of the output in the first version of Payment Transaction. For instance, if the output belongs to Alice, then the value of `outputGuard` equals Alice's address.
```
